src/libstore/build/derivation-goal.cc

```diff
@@ -3121,6 +3121,13 @@ void DerivationGoal::registerOutputs()
             newInfo0.references = refs.second;
             if (refs.first)
                 newInfo0.references.insert(newInfo0.path);
+            if (scratchPath != newInfo0.path) {
+                // Also rewrite the output path
+                rewriteOutput({{
+                    std::string(scratchPath.hashPart()),
+                    std::string(newInfo0.path.hashPart())
+                }});
```

Comment:
Sorry to contradict my own prior suggestion here, but I forgot about the two sets of `{..}`, maybe this would be better to visually distinguish input from output?
```suggestion
                rewriteOutput({
                    { std::string(scratchPath.hashPart()), std::string(newInfo0.path.hashPart()) }
                });
```

Reply:
I've actually reformated it in yet-another-way :stuck_out_tongue: (which makes me wish for https://github.com/NixOS/nix/issues/3697 to be a think), with an explicit call to the `StringMap` constructor because it's indeed to easy to forget what's going on otherwise

Reply:
Wait, how does that work, isn't `StringMap` the outer set of braces?

(Fine with the idea though!)

Reply:
> Wait, how does that work, isn't StringMap the outer set of braces?

I guess that's why people say that one shouldn't write code too late in the evening 🥴